### PR TITLE
Greytide Virus station event

### PR DESCRIPTION
**Pull request by slarticodefast**
## About the PR
Ports over the greytide virus station event from SS13 (see [tg wiki](https://tgstation13.org/wiki/Random_events#Greytide_Virus)).
It bolts open all airlocks and unlocks locks for a set of randomly chosen departments. The affected doors can be easily unbolted again by the AI, by using a door remote or by hacking the door. Unlike emagging this will not break the door's access reader.
The event has a severity from 1 to 3 which decides how many departments will be affected. Firelocks and doors with external access are ignored in order to not to space the station.

## Why / Balance
It is a classic from SS13, gives the AI something to do, and might lead to some fun prison break scenarios.

## Technical details
Added a new `TryOpenAndBolt` function to SharedDoorSystem to open and bolt a door. This works the same way emagging it does. However, it does not remove the access reader, so the door is fully repairable.

## Media

https://github.com/user-attachments/assets/c59fe3ab-d43e-47a2-98a1-c0b9c7a3a77c

Note: The bolt indicator light on the wirepanel currently does not update correctly until interacted with. This will be fixed in #33536

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
none

**Changelog**
:cl:
- add: Added the greytide virus station event. It will bolt open all doors in a few randomly chosen departments and unlock lockers with the corresponding access.

### CODE REVIEW

**Comment by hyphenationc** [comment]
Ooh
I'll look at this later
Thanks slarti barti :)

File: Content.Server/StationEvents/Events/GreytideVirusRule.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    protected override void Added(EntityUid uid, GreytideVirusRuleComponent virusComp, GameRuleComponent gameRule, GameRuleAddedEvent args)
+    {
+        if (!TryComp<StationEventComponent>(uid, out var stationEvent))
+            return;
+
+        // pick severity randomly from range if not specified otherwise
+        virusComp.Severity ??= virusComp.SeverityRange.Next(_random);
+        virusComp.Severity = Math.Min(virusComp.Severity.Value, virusComp.AccessGroups.Count);
+
+        stationEvent.StartAnnouncement = Loc.GetString("station-event-greytide-virus-start-announcement", ("severity", virusComp.Severity.Value));
```
Comment (lzk228) [inline Content.Server/StationEvents/Events/GreytideVirusRule.cs:36]: we should be getting rid of such hardcoded strings

File: Content.Shared/Doors/Systems/SharedDoorSystem.cs [hunk 2]
```diff
@@ -396,6 +396,25 @@ public void OnPartialOpen(EntityUid uid, DoorComponent? door = null)
         Dirty(uid, door);
 
     }
+
+    /// <summary>
+    /// Opens and then bolts a door.
+    /// Different from emagging this does not remove the access reader, so it can be repaired by simply unbolting the door.
+    /// </summary>
+    public bool TryOpenAndBolt(EntityUid uid, DoorComponent? door = null, AirlockComponent? airlock = null)
```
Comment (lzk228) [inline Content.Shared/Doors/Systems/SharedDoorSystem.cs:404]: i suppose you could realize emagdoor method via this new one

File: Content.Server/StationEvents/Events/GreytideVirusRule.cs (hunk 1 above)
Comment (slarticodefast) [reply Content.Server/StationEvents/Events/GreytideVirusRule.cs:36]: Move it into the component maybe?

File: Content.Shared/Doors/Systems/SharedDoorSystem.cs (hunk 2 above)
Comment (slarticodefast) [reply Content.Shared/Doors/Systems/SharedDoorSystem.cs:404]: No you can't. In the emag version the check for being unbolted and powered happens in a separate event, so that it does not consume a charge if it fails.

**Comment by Thatonestomf** [comment]
"The prisoners are already in the armory William. They will soon hit the streets angry and better armed then we are..."

Good pr and now it can affect other departments as well, very nice

**Comment by muburu** [comment]
minimum players should maybe be like 30, 10 is way too low and there's no guarantee there'll be an AI or all heads of staff at such lowpop

**Comment by slarticodefast** [comment]
> minimum players should maybe be like 30, 10 is way too low and there's no guarantee there'll be an AI or all heads of staff at such lowpop

Good point, I increased it to 25 to be more in line with other events.

**Comment by thebadman4662** [comment]
If severity 1 opens armory for passangers, does 3 spawn deathsquad gear in it while at it? 😆

**Comment by ScarKy0** [comment]
> If severity 1 opens armory for passangers, does 3 spawn deathsquad gear in it while at it? 😆

If i understand, severity is how many departments are affected at once

**Comment by RedBookcase** [comment]
Cool event, and i'm always in favor of anything to break up the monotony of vent creature spawns.

File: Content.Server/StationEvents/Events/GreytideVirusRule.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/GreytideVirusRule.cs:36]: I think it's fine as it's just the code-side of text that's customisable via localisation.

**Review by metalgearsloth — APPROVED** [review]

**Comment by cfif126** [comment]
can you please not delete branch as soon as possible, now i can't watch the video!

**Comment by slarticodefast** [comment]
> can you please not delete branch as soon as possible, now i can't watch the video!

The video is not in the branch though? I can still watch it, try refreshing the website

**Comment by slarticodefast** [comment]
> Can't download it though

That is an issue with your web browser then.
The download works perfectly fine for me.

**Comment by superjj18** [comment]
Legitimately in just two weeks when we get the next update

**Comment by Doc-Michael** [comment]
Alright i bring much needed **_feedback from LRP_**, and things that should have been brought up before it got merged, but to an important note : **Majority of Wizden servers operate on _Low RolePlay_, im saying this because a good MRP addition doesn't equal a good LRP one.**  Player culture is a factor that should not be overlooked.

- FEEDBACK

1. Doors should automatically UNBOLT after a given set of time, this will help if the AI is not being controlled by a connected player, or has been disconnected, and in cases where a Head of staff is not on the station or the remote has been stolen

2. It should really affect only 1 or 2 departments at the highest severity, a lot of the time, security is understaffed with 4 officers including HoS and Warden, leading to a huge inability to respond to callouts and threats, dealing with trespasses also requires officers and having 3 departments opened all at once is just overkill.

- THINGS THAT SHOULD HAVE BEEN SAID BEFORE THIS GOT MERGED

Let me begin with saying that its getting harder and harder to balance new features without making the game more in favour of antags/sec/greytide, while the game needs new content to not become stale, the content must be introduced in such a manner that offers new interactions as well as counterplay in order to be balanced to both parties, EMP to stunbaton, sunglasses to flashes etc. Some additions increase the skill requirement of the user while increasing the value of an addition if used correctly.

1. It severely affects the game balance while not providing much RP options (on LRP) as it claims to be, its Bolts open all the department doors (on Meta station if it affects engineering it affect external doors leading to the whole atmos getting vacuumed out(oh my god a bug report inside a feedback comment?)) going back, it only provides rewards greytiders (antagonists to a **SLIGHT** degree) while needlessly punishing actual departments which play a role in game progression, and an its current implementation it requires either Head of staff or AI intervention, where those might not always be available, or god forbid early day Lowpop happens, while that can be changed as i suggested to automatically unbolt after a set time (3 minutes? up to debate but not longer than 5) it still feels like a punishment to people who actually play a role other than a greytider, where on LRP the so called RP secenarios it provides are: What are you doing, get out of here, and then the tider and X department member fight to crit and either theft continues or the tider is dragged to med if its not med being affected.

2. A Heavily used argument on a lot of PRs is that they "Threw the game off balance" and im simply interested why was this, clear balance breaker was merged to begin with, ports over from ss13 were denied due to this exact reason but this one was added either way, while i appreciate the effort of the author that went into this PR, it simply does not meet the delicate balance that we had and threw it out the window in favour of "tider content", new content is good, but it must be implemented correctly to be a good addition, if it had for example a general console that you could use to decrease the effect of this event or prevent it, by forcing RD to rush to a X computer console to initiate a virus sweep before a 2 minute timer or less, offering some kind of counterplay.

3. It promotes self antagonism. Yes its a bit of a stretch to call trespassing a self-antag action, but it never ends on just trespassing, people when asked to leave will refuse to, leading to fights which most likely did not follow proper escalation, leading to round lasting grudges, decreasing the fun a person has in a round, you could say "just call sec", but on LRP sec is understaffed, busy dealing with actual antags or self-antags or 2 other departments which got affected, it simply overloads an already overloaded system.

TL:DR, it feels like it was a force merge of a not fully ironed out addition, it knocks the balance out of order while also being a huge annoyance to deal with, or being a huge deal on Low-Pop hours, while also rewarding borderline rule breaking behaviour.

Hope this proves helpful to either improve or redesign the current implementation of this PR.

**Comment by cfif126** [comment]
🥱

**Comment by Djungelskog2** [comment]
> Alright i bring much needed **_feedback from LRP_**, and things that should have been brought up before it got merged, but to an important note : **Majority of Wizden servers operate on _Low RolePlay_, im saying this because a good MRP addition doesn't equal a good LRP one.** Player culture is a factor that should not be overlooked.
> 
> * FEEDBACK
> 
> 1. Doors should automatically UNBOLT after a given set of time, this will help if the AI is not being controlled by a connected player, or has been disconnected, and in cases where a Head of staff is not on the station or the remote has been stolen
> 2. It should really affect only 1 or 2 departments at the highest severity, a lot of the time, security is understaffed with 4 officers including HoS and Warden, leading to a huge inability to respond to callouts and threats, dealing with trespasses also requires officers and having 3 departments opened all at once is just overkill.
> 
> * THINGS THAT SHOULD HAVE BEEN SAID BEFORE THIS GOT MERGED
> 
> Let me begin with saying that its getting harder and harder to balance new features without making the game more in favour of antags/sec/greytide, while the game needs new content to not become stale, the content must be introduced in such a manner that offers new interactions as well as counterplay in order to be balanced to both parties, EMP to stunbaton, sunglasses to flashes etc. Some additions increase the skill requirement of the user while increasing the value of an addition if used correctly.
> 
> 1. It severely affects the game balance while not providing much RP options (on LRP) as it claims to be, its Bolts open all the department doors (on Meta station if it affects engineering it affect external doors leading to the whole atmos getting vacuumed out(oh my god a bug report inside a feedback comment?)) going back, it only provides rewards greytiders (antagonists to a **SLIGHT** degree) while needlessly punishing actual departments which play a role in game progression, and an its current implementation it requires either Head of staff or AI intervention, where those might not always be available, or god forbid early day Lowpop happens, while that can be changed as i suggested to automatically unbolt after a set time (3 minutes? up to debate but not longer than 5) it still feels like a punishment to people who actually play a role other than a greytider, where on LRP the so called RP secenarios it provides are: What are you doing, get out of here, and then the tider and X department member fight to crit and either theft continues or the tider is dragged to med if its not med being affected.
> 2. A Heavily used argument on a lot of PRs is that they "Threw the game off balance" and im simply interested why was this, clear balance breaker was merged to begin with, ports over from ss13 were denied due to this exact reason but this one was added either way, while i appreciate the effort of the author that went into this PR, it simply does not meet the delicate balance that we had and threw it out the window in favour of "tider content", new content is good, but it must be implemented correctly to be a good addition, if it had for example a general console that you could use to decrease the effect of this event or prevent it, by forcing RD to rush to a X computer console to initiate a virus sweep before a 2 minute timer or less, offering some kind of counterplay.
> 3. It promotes self antagonism. Yes its a bit of a stretch to call trespassing a self-antag action, but it never ends on just trespassing, people when asked to leave will refuse to, leading to fights which most likely did not follow proper escalation, leading to round lasting grudges, decreasing the fun a person has in a round, you could say "just call sec", but on LRP sec is understaffed, busy dealing with actual antags or self-antags or 2 other departments which got affected, it simply overloads an already overloaded system.
> 
> TL:DR, it feels like it was a force merge of a not fully ironed out addition, it knocks the balance out of order while also being a huge annoyance to deal with, or being a huge deal on Low-Pop hours, while also rewarding borderline rule breaking behaviour.
> 
> Hope this proves helpful to either improve or redesign the current implementation of this PR.

Completely agree with all the points made from an admin perspective, it's self antag bait 

And it feels like some things weren't tested as lots of unintentional major things happened such as substation rooms, external doors, ect

**Comment by deltanedas** [comment]
you are overlooking that engis can unbolt doors, but overall i agree

**Comment by Djungelskog2** [comment]
as if engis actually do their job :Godo:

**Comment by ScarKy0** [comment]
Heads also have door remotes to like instantly unbolt things in their departments
